### ISSUE TITLE
Is there a way to pass an EventCallback as a CascadingValue without using internal framework use methods?

### ISSUE THREAD

**Issue by mrlife** [dotnet/aspnetcore#24883]
# Issue description

Some methods are described with "For internal framework use only."

I have been using `EventCallback.Factory.Create(Object, EventCallback)` to pass an `EventCallback` as a `CascadingValue`. This allows me to pass the EventCallback down a few levels without having to pass it explicitly through each razor component between the one with the callback function and the one that will call it.

I recently saw in the [documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.eventcallbackfactory.create?view=aspnetcore-3.1#Microsoft_AspNetCore_Components_EventCallbackFactory_Create_System_Object_Microsoft_AspNetCore_Components_EventCallback_) it is described by the following:

> Returns the provided callback. For internal framework use only.

I assume this means it should not be used or that it can disappear from a future release (with or without notice).

I've tried using the other `Create` methods in `EventCallback.Factory`/`EventCallbackFactory`, but haven't been able to accomplish the same thing without compile errors.

cc: @guardrex

**Comment by blowdart** [comment]
Both. Or it's behaviour may change. Basically don't use it.

Instead raise a feature request describing what you're doing and how it solves it and lobby for a public feature for your user case.

**Comment by mrlife** [comment]
Thanks, @blowdart. Possible to answer a quick question here? I think the following would solve this for me: is there a way to pass an EventCallback as a `CascadingValue` without using `EventCallback.Factory.Create(Object, EventCallback)`?

**Comment by blowdart** [comment]
Ok that I can't comment on. Let me edit the title so it's better and someone might pick it up.

**Comment by javiercn** [comment]
@mrlife thanks for contacting us.

Can you tell us more about what you are trying to accomplish? We might be able to suggest alternatives. The statement in the documentation is to warn users that we might change those APIs in the future as we need and that they'll have to react to the changes.

If you take a dependency on them, you'll have to react if we change them.

**Comment by mrlife** [comment]
@javiercn It is simpler and uses less boilerplate code sprinkled through component hierarchy (purpose of `CascadingValue`). 

Using `EventCallback.Factory.Create(Object, EventCallback)` allows me to pass the EventCallback down 4-5 levels in a component hierarchy as a `CascadingValue` \ `CascadingParameter` without having to pass it explicitly through each razor component between the one with the callback function (ancestor component) and the one that will call it (descendant component).

**Comment by captainsafia** [comment]
@mrlife Can you try using the public `EventCallback` constructor?

**Comment by mrlife** [comment]
@captainsafia I have tried a few; which one do you recommend?

**Comment by SteveSandersonMS** [comment]
> CS1503: Argument 2: cannot convert from 'method group' to 'MulticastDelegate'

You can explicitly cast to `MulticastDelegate`, or cast to a different delegate type such as `Action`.

**Comment by mrlife** [comment]
@SteveSandersonMS Thank you, I tried your suggestion but ran into an issue.

I changed my current code:

```
var callback = EventCallback.Factory.Create<MyClass>(this, MyMethod)
```

to:

```
var callback = new EventCallback<MyClass>(this, (MulticastDelegate)MyMethod);
```

and see this error:

> Error CS0030: Cannot convert type 'method' to 'MulticastDelegate'

**Comment by SteveSandersonMS** [comment]
This is a matter of C# syntax in general, not specific to Blazor. You can't cast directly from a methodgroup (method name) to a `MulticastDelegate` because the compiler doesn't know what signature delegate you want. Instead you have to cast to a specific delegate type first. For example:

```cs
var callback = new EventCallback<MyClass>(this, (MulticastDelegate)(Action)MyMethod);
```

Or in this case, since `Action` can be converted implicitly to `MulticastDelegate`, you could simplify it to:

```cs
var callback = new EventCallback<MyClass>(this, (Action)MyMethod);
```

**Comment by mrlife** [comment]
@SteveSandersonMS Thank you very much! I got it working with a slight alteration to the second line (due to `MyMethod` taking a parameter of type `MyClass`):

```c#
var callback = new EventCallback<MyClass>(this, (Action<MyClass>)MyMethod);
```

I have another one that has a method that returns `Task` and takes no parameters, so I used the following for that one. Putting this here for anyone else who might be unsure about how to handle this case.

```c#
var callback = new EventCallback<MyClass>(this, (Func<Task>)MyMethod);
```

**Comment by mrlife** [comment]
@SteveSandersonMS Thank you very much! I got it working with a slight alteration to the second line (due to `MyMethod` taking a parameter of type `MyClass`):

```c#
var callback = new EventCallback<MyClass>(this, (Action<MyClass>)MyMethod);
```

I have another one that has a method that returns `Task` and takes no parameters, so I used the following for that one. Putting this here for anyone else who might be unsure about how to handle this case.

```c#
var callback = new EventCallback(this, (Func<Task>)MyMethod);
```
